### PR TITLE
refactor: introduce QueryKind.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

`query_kind` was originally only used in query_log, so it is just a `String`.
later we found it useful as a lightweight context in some scenarios,  
let us formalize it as an `Enum` for safety.

`common-sql` already depends on `common-catalog`, so no extra deps is introduced.
 
Summary about this PR

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12965)

### CODE REVIEW

File: src/query/sql/src/planner/plans/plan.rs [hunk 1]
```diff
@@ -302,123 +304,121 @@ pub enum RewriteKind {
     Call,
 }
 
-impl Display for Plan {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+impl Plan {
+    pub fn kind(&self) -> QueryKind {
```
Comment (sundy-li) [inline src/query/sql/src/planner/plans/plan.rs:308]: Notice that QueryKind is not accurate.

Some statements have been rewritten to Query like `ShowTables`

Comment (sundy-li) [reply src/query/sql/src/planner/plans/plan.rs:308]: Why not match the AST statement ?

Comment (sundy-li) [reply src/query/sql/src/planner/plans/plan.rs:308]: we can cache the  AST statement  in context.


```
pub(in crate::sessions) running_query_stmt: Statment,
```

Comment (youngsofun) [reply src/query/sql/src/planner/plans/plan.rs:308]: currently only used for copy.

Comment (youngsofun) [reply src/query/sql/src/planner/plans/plan.rs:308]: nn

> we can cache the AST statement in context.
> 
> ```
> pub(in crate::sessions) running_query_stmt: Arc<RwLock<Option<Statement>>>,,
> ```

need to be serialized and sent to other query nodes.

Comment (youngsofun) [reply src/query/sql/src/planner/plans/plan.rs:308]: also comparation is faster

Comment (youngsofun) [reply src/query/sql/src/planner/plans/plan.rs:308]: > nn
> 
> > we can cache the AST statement in context.
> > ```
> > pub(in crate::sessions) running_query_stmt: Arc<RwLock<Option<Statement>>>,,
> > ```
> 
> need to be serialized and sent to other query nodes.

Statement not impl Serialize for now, even if it does, it it too big for current usage.

File: src/query/storages/parquet/src/parquet2/parquet_table/partition.rs [hunk 2]
```diff
@@ -159,7 +160,7 @@ impl Parquet2Table {
                 &file_locations,
                 ctx.get_settings().get_max_threads()? as usize,
                 &ctx.get_copy_status(),
-                ctx.get_query_kind().eq_ignore_ascii_case("copy"),
+                ctx.get_query_kind() == QueryKind::Copy,
```
Comment (ariesdevil) [inline src/query/storages/parquet/src/parquet2/parquet_table/partition.rs:163]: The convention for rust is to compare enum with `matches!`

File: src/query/sql/src/planner/plans/plan.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/sql/src/planner/plans/plan.rs:308]: Ok, let's respect `RewriteKind` in query plan.

Comment (youngsofun) [reply src/query/sql/src/planner/plans/plan.rs:308]: QueryKind is from Plan,  those stmt in RewriteKind are already rewritten to Query at this point.

File: src/query/storages/parquet/src/parquet2/parquet_table/partition.rs (hunk 2 above)
Comment (youngsofun) [reply src/query/storages/parquet/src/parquet2/parquet_table/partition.rs]: done

**Comment by youngsofun** [comment]
@sundy-li  plz review again https://github.com/datafuselabs/databend/pull/12965/commits/1d4312f6c3b9163810b78fc3592f3f382a3ff569

**Review by sundy-li — APPROVED** [review]

**Pull request merged**